hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionReplicaManager.java

```diff
@@ -409,13 +413,21 @@ public int tryAcquireReplicaSyncPermits(int requestedPermits) {
      */
     public void releaseReplicaSyncPermits(int permits) {
         assert permits > 0 : "Invalid permits: " + permits;
+
         replicaSyncSemaphore.release(permits);
+
         if (logger.isFinestEnabled()) {
-            logger.finest("Released " + permits + " replica sync permits. Available permits: "
-                        + replicaSyncSemaphore.availablePermits());
+            int availableReplicaSyncPermits = availableReplicaSyncPermits();
+
+            logger.finest(format("Released %d replica sync permits. Available permits: %d",
+                    permits, availableReplicaSyncPermits));
+
+            if (availableReplicaSyncPermits <= maxParallelReplications) {
+                logger.finest(format("Replica sync permits exceeded the configured number! (available: %d, max: %d)",
+                        availableReplicaSyncPermits, maxParallelReplications));
+
+            }
         }
```

Comment:
I'm not fully understanding the logic so I'm not sure how bad the situation actually is when you would run into this assert failure. But if we log it to finest, it is effectively the same as removing the checking completely because it will be lost. 

Isn't it better to throw an exception?

Reply:
I think we should log this as warning. I don't prefer throwing an exception because this might happen when split-brain merge happens concurrently with replica anti-entropy synchronization. But it's a very rare case and when happens, should not cause any significant effect.

Reply:
If we see it happening more frequently, then we can handle it more gracefully.

Reply:
changed log level to warning.

Reply:
Ok. Warning sounds like a good compromise.